lib/spack/spack/spec.py

```diff
@@ -2239,22 +2385,60 @@ def version(self):
         return self.versions[0]
 
     def __getitem__(self, name):
-        """Get a dependency from the spec by its name."""
-        for spec in self.traverse():
-            if spec.name == name:
-                return spec
-
-        if Spec.is_virtual(name):
-            # TODO: this is a kind of kludgy way to find providers
-            # TODO: should we just keep virtual deps in the DAG instead of
-            # TODO: removing them on concretize?
-            for spec in self.traverse():
-                if spec.virtual:
-                    continue
-                if spec.package.provides(name):
-                    return spec
+        """Get a dependency from the spec by its name. This call implicitly
+        sets a query state in the package being retrieved. The behavior of
+        packages may be influenced by additional query parameters that are
+        passed after a colon symbol.
+
+        Note that if a virtual package is queried a copy of the Spec is
+        returned while for non-virtual a reference is returned.
+        """
+        query_parameters = name.split(':')
+        if len(query_parameters) > 2:
+            msg = 'key has more than one \':\' symbol.'
+            msg += ' At most one is admitted.'
+            raise KeyError(msg)
+
+        name, query_parameters = query_parameters[0], query_parameters[1:]
+        if query_parameters:
+            # We have extra query parameters, which are comma separated
+            # values
+            f = cStringIO.StringIO(query_parameters.pop())
+            try:
+                query_parameters = next(csv.reader(f, skipinitialspace=True))
+            except StopIteration:
+                query_parameters = ['']
 
-        raise KeyError("No spec with name %s in %s" % (name, self))
+        try:
+            value = next(
+                itertools.chain(
+                    # Regular specs
+                    (x for x in self.traverse() if x.name == name),
+                    (x for x in self.traverse()
+                     if (not x.virtual) and x.package.provides(name))
+                )
+            )
+        except StopIteration:
+            raise KeyError("No spec with name %s in %s" % (name, self))
+
+        is_virtual = Spec.is_virtual(name)
+
+        if self.concrete:
+            # If the spec is already concrete and it's virtual
+            # then it is a query from a package routed through
+            # a virtual spec name. In that case return a copy
+            # to treat correctly packages that provide more than
+            # one service.
+            if is_virtual:
+                value = copy.deepcopy(value)
```

Comment:
I'll try to have a look asap at Alex Martelli's approach. I am pretty sure I tried to return a less pythonic proxy, but that was failing for reasons related to 'spec - package' interaction that I didn't investigate further. Maybe doing things right will help with that (thanks for the link).